From 5d07f0d295302ae3f60bdeba695ff3d9d5a5ea3d Mon Sep 17 00:00:00 2001
From: Jike Song <jike.song@intel.com>
Date: Mon, 11 May 2015 08:11:06 +0800
Subject: [PATCH 4/8] drm: use a dirty workaround to remove dependency between
 drm and i915

The irq mediation logic is going anyway. It's safe to remove this
workaround when host i915 becomes GVT-g aware.

Signed-off-by: Jike Song <jike.song@intel.com>
---
 drivers/gpu/drm/drm_irq.c       | 18 +++++++++---------
 drivers/gpu/drm/i915/i915_dma.c |  8 ++++++++
 2 files changed, 17 insertions(+), 9 deletions(-)

diff --git a/drivers/gpu/drm/drm_irq.c b/drivers/gpu/drm/drm_irq.c
index e291af2..3ec699e 100644
--- a/drivers/gpu/drm/drm_irq.c
+++ b/drivers/gpu/drm/drm_irq.c
@@ -43,10 +43,10 @@
 #include <linux/export.h>
 
 #ifdef CONFIG_I915_VGT
-extern bool i915_host_mediate __read_mostly;
-extern bool vgt_can_process_irq(void);
-extern bool vgt_can_process_timer(void *timer);
-extern void vgt_new_delay_event_timer(void *timer);
+bool (*tmp_vgt_can_process_timer)(void *timer) = NULL;
+void (*tmp_vgt_new_delay_event_timer)(void *timer) = NULL;
+EXPORT_SYMBOL(tmp_vgt_can_process_timer);
+EXPORT_SYMBOL(tmp_vgt_new_delay_event_timer);
 #endif
 
 /* Access macro for slots in vblank timestamp ringbuffer. */
@@ -260,9 +260,9 @@ static void vblank_disable_fn(unsigned long arg)
 		return;
 
 #ifdef CONFIG_I915_VGT
-	if (i915_host_mediate)
-		if (!vgt_can_process_timer(&vblank->disable_timer))
-			return;
+	if (tmp_vgt_new_delay_event_timer &&
+			!tmp_vgt_can_process_timer(&vblank->disable_timer))
+		return;
 #endif
 
 	spin_lock_irqsave(&dev->vbl_lock, irqflags);
@@ -336,8 +336,8 @@ int drm_vblank_init(struct drm_device *dev, int num_crtcs)
 		setup_timer(&vblank->disable_timer, vblank_disable_fn,
 			    (unsigned long)vblank);
 #ifdef CONFIG_I915_VGT
-		if (i915_host_mediate)
-			vgt_new_delay_event_timer(&vblank->disable_timer);
+		if (tmp_vgt_new_delay_event_timer)
+			tmp_vgt_new_delay_event_timer(&vblank->disable_timer);
 #endif
 	}
 
diff --git a/drivers/gpu/drm/i915/i915_dma.c b/drivers/gpu/drm/i915/i915_dma.c
index d94897c..53d9551 100644
--- a/drivers/gpu/drm/i915/i915_dma.c
+++ b/drivers/gpu/drm/i915/i915_dma.c
@@ -706,6 +706,14 @@ int i915_driver_load(struct drm_device *dev, unsigned long flags)
 		i915_host_mediate = true;
 	printk("i915_start_vgt: %s\n", i915_host_mediate ? "success" : "fail");
 
+	if (i915_host_mediate) {
+		extern bool (*tmp_vgt_can_process_timer)(void *timer);
+		extern void (*tmp_vgt_new_delay_event_timer)(void *timer);
+
+		tmp_vgt_new_delay_event_timer = vgt_new_delay_event_timer;
+		tmp_vgt_can_process_timer = vgt_can_process_timer;
+	}
+
 	i915_check_vgt(dev_priv);
 	if (USES_VGT(dev))
 		i915.enable_fbc = 0;
-- 
1.9.1

